#### <b> Dataset </b>

- Download the `dataset_1 and dataset_2` from Course Resources and upload the datasets to the lab

### **Import libraries**

- Pandas is a high-level data manipulation tool

- NumPy is used for working with multidimensional arrays

In [2]:
import pandas as pd
import numpy as np

### **Load the first dataset**

In [2]:
dataset_1 = pd.read_csv('dataset_1.csv')
dataset_1.head()

instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
0        1  01-01-2011       1   0     1   0    False        6           1   
1        2  01-01-2011       1   0     1   1    False        6           1   
2        3  01-01-2011       1   0     1   2    False        6           1   
3        4  01-01-2011       1   0     1   3    False        6           1   
4        5  01-01-2011       1   0     1   4    False        6           1   

   temp  
0  0.24  
1  0.22  
2  0.22  
3  0.24  
4  0.24

In [3]:
type(dataset_1)

pandas.core.frame.DataFrame

#### **Observations:**

- The result shows that the dataset is DataFrame

- DataFrame is a tabular structure consisting of rows and columns

In [4]:
# Shape of the dataset
dataset_1.shape

(610, 10)

#### **Observation**:

- The `dataset_1` has 610 rows and 10 columns.

### **Load the second dataset**

- Use the function carefully since it is an excel file

In [5]:
dataset_2 = pd.read_excel('dataset_2.xlsx')
dataset_2.head()

Unnamed: 0  instant   atemp   hum  windspeed  casual  registered  cnt
0           0        1  0.2879  0.81        0.0       3          13   16
1           1        2  0.2727  0.80        0.0       8          32   40
2           2        3  0.2727  0.80        0.0       5          27   32
3           3        4  0.2879  0.75        0.0       3          10   13
4           4        5  0.2879  0.75        0.0       0           1    1

#### **Observation:**
- We can see a column named `unnamed:0`, which is not in the data dictionary. Let's remove it.

### **Drop the column**

In [6]:
dataset_2.shape

(610, 8)

In [7]:
dataset_2.drop(['Unnamed: 0'], axis=1, inplace = True)
dataset_2.shape

# df_name.drop(['name_of_the column'], col axis = 1, changes will be made in the df)

(610, 7)

#### **Observation:**

- We had 8 columns before the drop.

- When we check the shape of the file after the drop, we see that the column `Unnamed: 0` has been dropped

In [8]:
dataset_2.head()

instant   atemp   hum  windspeed  casual  registered  cnt
0        1  0.2879  0.81        0.0       3          13   16
1        2  0.2727  0.80        0.0       8          32   40
2        3  0.2727  0.80        0.0       5          27   32
3        4  0.2879  0.75        0.0       3          10   13
4        5  0.2879  0.75        0.0       0           1    1

### **Merge the datasets**

- We have two datasets. They are dataset_1 and dataset_2

- As both datasets have one common column 'instant', let's merge the datasets on that column

- We are going to save the resultant data inside the **combined_data** as shown below

In [9]:
display (dataset_1.head(2))
display (dataset_2.head(2))

instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
0        1  01-01-2011       1   0     1   0    False        6           1   
1        2  01-01-2011       1   0     1   1    False        6           1   

   temp  
0  0.24  
1  0.22

instant   atemp   hum  windspeed  casual  registered  cnt
0        1  0.2879  0.81        0.0       3          13   16
1        2  0.2727  0.80        0.0       8          32   40

In [10]:
print (dataset_1.columns)
print ()
print (dataset_2.columns)

Index(['instant', 'dteday', 'season', 'yr', 'mnth', 'hr', 'holiday', 'weekday',
       'weathersit', 'temp'],
      dtype='object')

Index(['instant', 'atemp', 'hum', 'windspeed', 'casual', 'registered', 'cnt'], dtype='object')


In [11]:
print (dataset_1.shape)
print ()
print (dataset_2.shape)

(610, 10)

(610, 7)


In [12]:
combined_data = pd.merge(dataset_1, dataset_2, on='instant')
combined_data.shape

(610, 16)

#### **Observation:**

- The shape of the combined_data has 610 rows and 16 columns 

In [13]:
combined_data.head()

instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
0        1  01-01-2011       1   0     1   0    False        6           1   
1        2  01-01-2011       1   0     1   1    False        6           1   
2        3  01-01-2011       1   0     1   2    False        6           1   
3        4  01-01-2011       1   0     1   3    False        6           1   
4        5  01-01-2011       1   0     1   4    False        6           1   

   temp   atemp   hum  windspeed  casual  registered  cnt  
0  0.24  0.2879  0.81        0.0       3          13   16  
1  0.22  0.2727  0.80        0.0       8          32   40  
2  0.22  0.2727  0.80        0.0       5          27   32  
3  0.24  0.2879  0.75        0.0       3          10   13  
4  0.24  0.2879  0.75        0.0       0           1    1

In [14]:
combined_data.tail()

instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
605      606  28-01-2011       1   0     1  11    False        5           3   
606      607  28-01-2011       1   0     1  12    False        5           3   
607      608  28-01-2011       1   0     1  13    False        5           3   
608      609  28-01-2011       1   0     1  14    False        5           3   
609      610  28-01-2011       1   0     1  15    False        5           2   

     temp   atemp   hum  windspeed  casual  registered  cnt  
605  0.18  0.2121  0.93     0.1045       0          30   30  
606  0.18  0.2121  0.93     0.1045       1          28   29  
607  0.18  0.2121  0.93     0.1045       0          31   31  
608  0.22  0.2727  0.80     0.0000       2          36   38  
609  0.20  0.2576  0.86     0.0000       1          40   41

### **Import the dataset_3**

In [15]:
dataset_3 = pd.read_csv('dataset_3.csv')
dataset_3.shape

(390, 16)

In [16]:
dataset_3.head()

instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
0      620  29-01-2011       1   0     1   1    False        6           1   
1      621  29-01-2011       1   0     1   2    False        6           1   
2      622  29-01-2011       1   0     1   3    False        6           1   
3      623  29-01-2011       1   0     1   4    False        6           1   
4      624  29-01-2011       1   0     1   6    False        6           1   

   temp   atemp   hum  windspeed  casual  registered  cnt  
0  0.22  0.2273  0.64     0.1940       0          20   20  
1  0.22  0.2273  0.64     0.1642       0          15   15  
2  0.20  0.2121  0.64     0.1343       3           5    8  
3  0.16  0.1818  0.69     0.1045       1           2    3  
4  0.16  0.1818  0.64     0.1343       0           2    2

In [17]:
dataset_3.tail(15)

instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
375      995  14-02-2011       1   0     2   2    False        1           1   
376      996  14-02-2011       1   0     2   3    False        1           1   
377      997  14-02-2011       1   0     2   4    False        1           1   
378      998  14-02-2011       1   0     2   5    False        1           1   
379      999  14-02-2011       1   0     2   6    False        1           1   
380     1000  14-02-2011       1   0     2   7    False        1           1   
381      611  28-01-2011       1   0     1  16    False        5           1   
382      612  28-01-2011       1   0     1  17    False        5           1   
383      613  28-01-2011       1   0     1  18    False        5           1   
384      614  28-01-2011       1   0     1  19    False        5           2   
385      615  28-01-2011       1   0     1  20    False        5           2   
386      616  28-01-2011       1   0     1  21    False        5           2   
387      617  28-01-2011       1   0     1  22    False        5           1   
388      618  28-01-2011       1   0     1  23    False        5           1   
389      619  29-01-2011       1   0     1   0    False        6           1   

     temp   atemp   hum  windspeed  casual  registered  cnt  
375  0.36  0.3333  0.40     0.2985       0           2    2  
376  0.34  0.3182  0.46     0.2239       1           1    2  
377  0.32  0.3030  0.53     0.2836       0           2    2  
378  0.32  0.3030  0.53     0.2836       0           3    3  
379  0.34  0.3030  0.46     0.2985       1          25   26  
380  0.34  0.3030  0.46     0.2985       2          96   98  
381  0.22  0.2727  0.80     0.0000      10          70   80  
382  0.24  0.2424  0.75     0.1343       2         147  149  
383  0.24  0.2273  0.75     0.1940       2         107  109  
384  0.24  0.2424  0.75     0.1343       5          84   89  
385  0.24  0.2273  0.70     0.1940       1          61   62  
386  0.22  0.2273  0.75     0.1343       1          57   58  
387  0.24  0.2121  0.65     0.3582       0          26   26  
388  0.24  0.2273  0.60     0.2239       1          22   23  
389  0.22  0.1970  0.64     0.3582       2          26   28

#### **Observation:**

- Here, we can see that the rows are not sorted well according to the `instant` number. Let's resolve it.

### **Sort values of a column**

- To sort the values per our will, we use the **`sort_values`** function and in the square brackets, we specify the name of the column by which we want to sort, as shown below

In [18]:
dataset_3 = dataset_3.sort_values(by=['instant'])  # default will be ascending ; ascending=True
dataset_3.head(2)

# dataset_3.sort_values(by=['instant'], inplace = True) 

instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
381      611  28-01-2011       1   0     1  16    False        5           1   
382      612  28-01-2011       1   0     1  17    False        5           1   

     temp   atemp   hum  windspeed  casual  registered  cnt  
381  0.22  0.2727  0.80     0.0000      10          70   80  
382  0.24  0.2424  0.75     0.1343       2         147  149

In [19]:
dataset_3.tail(2)

instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
379      999  14-02-2011       1   0     2   6    False        1           1   
380     1000  14-02-2011       1   0     2   7    False        1           1   

     temp  atemp   hum  windspeed  casual  registered  cnt  
379  0.34  0.303  0.46     0.2985       1          25   26  
380  0.34  0.303  0.46     0.2985       2          96   98

In [20]:
dataset_3.shape

(390, 16)

### **Concatenate the `combine_data` with `dataset_3`**

- Let's concatenate both DataFrame combined_data and dataset_3 into a single DataFrame using the **`concat`** function, as shown below
- Store the final DataFrame inside the `final_data` variable

In [21]:
final_data = pd.concat([combined_data, dataset_3])
final_data.shape

(1000, 16)

#### **Observation:**

- Now, the `final_data` has 1000 rows and 16 columns

In [22]:
final_data.head()

instant      dteday  season  yr  mnth  hr  holiday  weekday  weathersit  \
0        1  01-01-2011       1   0     1   0    False        6           1   
1        2  01-01-2011       1   0     1   1    False        6           1   
2        3  01-01-2011       1   0     1   2    False        6           1   
3        4  01-01-2011       1   0     1   3    False        6           1   
4        5  01-01-2011       1   0     1   4    False        6           1   

   temp   atemp   hum  windspeed  casual  registered  cnt  
0  0.24  0.2879  0.81        0.0       3          13   16  
1  0.22  0.2727  0.80        0.0       8          32   40  
2  0.22  0.2727  0.80        0.0       5          27   32  
3  0.24  0.2879  0.75        0.0       3          10   13  
4  0.24  0.2879  0.75        0.0       0           1    1

#### **Renaming the columns names**

In [23]:
final_data =  final_data.rename(columns={'dteday': 'date', 'yr': 'year', 'mnth':'month','hr':'hour',
                                        'weathersit':'weather', 'hum':'humidity', 'cnt':'count'})
final_data.head(2)

instant        date  season  year  month  hour  holiday  weekday  weather  \
0        1  01-01-2011       1     0      1     0    False        6        1   
1        2  01-01-2011       1     0      1     1    False        6        1   

   temp   atemp  humidity  windspeed  casual  registered  count  
0  0.24  0.2879      0.81        0.0       3          13     16  
1  0.22  0.2727      0.80        0.0       8          32     40

## **Data Dictionary**

**Attribute Information:**

* date = date of the ride  
* season -  1 = spring, 2 = summer, 3 = fall, 4 = winter 
* yr - 0 == 2010, 1 == 2011 
* month - 1-12
* holiday - whether the day is considered a holiday
* weekday - day of the week
* workingday - whether the day is neither a weekend nor holiday
* weather - 1: Clear, Few clouds, Partly cloudy, Partly cloudy
            2: Mist + Cloudy, Mist + Broken clouds, Mist + Few clouds, Mist
            3: Light Snow, Light Rain + Thunderstorm + Scattered clouds, Light Rain + Scattered clouds
            4: Heavy Rain + Ice Pallets + Thunderstorm + Mist, Snow + Fog 
* temp - temperature in Celsius
* atemp - "feels like" temperature in Celsius
* humidity - relative humidity
* windspeed - wind speed
* casual - number of non-registered user rentals initiated
* registered - number of registered user rentals initiated
* count - number of total rentals

### **Data types of different column values**

In [25]:
final_data.dtypes

instant         int64
date           object
season          int64
year            int64
month           int64
hour            int64
holiday          bool
weekday         int64
weather         int64
temp          float64
atemp         float64
humidity      float64
windspeed     float64
casual          int64
registered      int64
count           int64
dtype: object

#### **Observations:**

- We can see that the majority of our data columns are of type int64. They are therefore 64-bit integers. Some of the columns are of the type float64, which implies that they have decimals in them. However, only the date column has an object type, indicating that it contains strings.

# Missing Values

### **Check for null values**

- Execute the given command to check the unknown values in the DataFrame


In [26]:
final_data.isna()

instant   date  season   year  month   hour  holiday  weekday  weather  \
0      False  False   False  False  False  False    False    False    False   
1      False  False   False  False  False  False    False    False    False   
2      False  False   False  False  False  False    False    False    False   
3      False  False   False  False  False  False    False    False    False   
4      False  False   False  False  False  False    False    False    False   
..       ...    ...     ...    ...    ...    ...      ...      ...      ...   
376    False  False   False  False  False  False    False    False    False   
377    False  False   False  False  False  False    False    False    False   
378    False  False   False  False  False  False    False    False    False   
379    False  False   False  False  False  False    False    False    False   
380    False  False   False  False  False  False    False    False    False   

      temp  atemp  humidity  windspeed  casual  registered  count  
0    False  False     False      False   False       False  False  
1    False  False     False      False   False       False  False  
2    False  False     False      False   False       False  False  
3    False  False     False      False   False       False  False  
4    False  False     False      False   False       False  False  
..     ...    ...       ...        ...     ...         ...    ...  
376  False  False     False      False   False       False  False  
377  False  False     False      False   False       False  False  
378  False  False     False      False   False       False  False  
379  False  False     False      False   False       False  False  
380  False  False     False      False   False       False  False  

[1000 rows x 16 columns]

#### **Observation:**

- The **`isna()`** function returns DataFrame of Boolean values that are True for null values  

- In a huge dataset, the code given above is not going to help

- We do not get enough idea of the null values by looking at the given tabular dataset

- The next line of code is more convenient in this case.

In [27]:
final_data.isna().sum() # col sum; True = 1, False = 0

instant        0
date           0
season         0
year           0
month          0
hour           0
holiday        0
weekday        0
weather        0
temp           0
atemp         11
humidity       0
windspeed      0
casual         0
registered     0
count          0
dtype: int64

#### **Observations:**

- The `isna().sum()` function provides a clear picture of the number of null values in a DataFrame

-  In the given result, we can see that the `atemp` column has 11 null values

### **Let's check the percentage of the rows with missing value**

- We are performing this operation to determine whether the NA value rows can be dropped off or not so that we cannot deviate from our desired model

In [33]:
percentage_of_missing_values = (final_data['atemp'].isna().sum()/final_data.shape[0])*100
percentage_of_missing_values

1.0999999999999999

#### **Observations:**

- We divide the number of null values by the shape of the DataFrame to get the percentage of missing values.

- Since the percentage is 1, it is very less. Usually, the industry practice allows us to drop rows up to 30%. So, we can drop the rows with missing values.

### **Drop the rows with missing values**

- We will use the `dropna` function to drop the null value rows

In [34]:
print (final_data.shape)
final_data.dropna(inplace = True)
print (final_data.shape)

(1000, 16)
(989, 16)


#### **Observations:**

- We can see that the shape of the DataFrame reduced to 989 from 1000. It shows that the missing value rows have been wiped off.

- In further lessons of this course, we'll see different methods to treat missing values.

### **Now, let's again check the missing value count after the drop**

In [35]:
final_data.isna().sum()

instant       0
date          0
season        0
year          0
month         0
hour          0
holiday       0
weekday       0
weather       0
temp          0
atemp         0
humidity      0
windspeed     0
casual        0
registered    0
count         0
dtype: int64

## **Perform sanity checks on the dataset**

- It verifies the logical correctness of the data points

In [36]:
final_data.head()

instant        date  season  year  month  hour  holiday  weekday  weather  \
0        1  01-01-2011       1     0      1     0    False        6        1   
1        2  01-01-2011       1     0      1     1    False        6        1   
2        3  01-01-2011       1     0      1     2    False        6        1   
3        4  01-01-2011       1     0      1     3    False        6        1   
4        5  01-01-2011       1     0      1     4    False        6        1   

   temp   atemp  humidity  windspeed  casual  registered  count  
0  0.24  0.2879      0.81        0.0       3          13     16  
1  0.22  0.2727      0.80        0.0       8          32     40  
2  0.22  0.2727      0.80        0.0       5          27     32  
3  0.24  0.2879      0.75        0.0       3          10     13  
4  0.24  0.2879      0.75        0.0       0           1      1

### **Check if `casual` + `registered` is always equal to `count`**

In [38]:
final_data['casual'] + final_data['registered'] - final_data['count']
np.sum(final_data['casual'] + final_data['registered'] - final_data['count'])

0

### **Month values should be in the range of 1-12**

- We will use the `unique()` function to find the elements of an array

In [41]:
final_data.month.unique() # we will need to investigate

array([1, 2])

#### **Observations:**

- The data is truncated and hence, we have infomration only about 2 months

### **Hour should be in the range of 1-24**

In [42]:
final_data.hour.unique()

array([ 0,  1,  2,  3,  4,  5,  6,  7,  9, 10, 11, 12, 13, 14, 15, 16, 18,
       19, 20, 21, 22,  8, 17, 23])

### **For Date Manipulation**

In [3]:
import datetime

dti = pd.to_datetime(["1/1/2018", np.datetime64("2018-01-01"), datetime.datetime(2018, 1, 1)])
dti

DatetimeIndex(['2018-01-01', '2018-01-01', '2018-01-01'], dtype='datetime64[ns]', freq=None)

#### Difference between dataframe_name.isna( ) and dataframe_name.isnull( ) ?
- These two DataFrame methods do exactly the same thing! Even their docs are identical. You can even confirm this in pandas' code.
But why have two methods with different names do the same thing?

This is because pandas' DataFrames are based on R's DataFrames. In R na and null are two separate things. Read this post for more information.

However, in python, pandas is built on top of numpy, which has neither na nor null values. Instead numpy has NaN values (which stands for "Not a Number"). Consequently, pandas also uses NaN values.
In short

    To detect NaN values numpy uses np.isnan().

    To detect NaN values pandas uses either .isna() or .isnull().
    The NaN values are inherited from the fact that pandas is built on top of numpy, while the two functions' names originate from R's DataFrames, whose structure and functionality pandas tried to mimic.
